File: protocol/src/test/typescript/zfoo/packet/ObjectA.ts
import ObjectB from './ObjectB';

// @author godotg
// @version 3.0
class ObjectA {

    a: number = 0;
    m: Map<number, string> | null = null;
    objectB: ObjectB | null = null;

    protocolId(): number {
        return 102;
    }

    static write(buffer: any, packet: ObjectA | null) {
        if (buffer.writePacketFlag(packet)) {
            return;
        }
        if (packet === null) {
            return;
        }

        buffer.writeInt(packet.a);
        buffer.writeIntStringMap(packet.m);
        buffer.writePacket(packet.objectB, 103);
    }

    static read(buffer: any): ObjectA | null {
        if (!buffer.readBoolean()) {
            return null;
        }
        const packet = new ObjectA();
        const result0 = buffer.readInt();
        packet.a = result0;
        const map1 = buffer.readIntStringMap();
        packet.m = map1;
        const result2 = buffer.readPacket(103);
        packet.objectB = result2;
        return packet;
    }
}

export default ObjectA;

File: protocol/src/test/lua/LuaProtocol/Packet/ObjectA.lua
-- @author godotg

local ObjectA = {}

function ObjectA:new(a, m, objectB)
    local obj = {
        a = a, -- int
        m = m, -- java.util.Map<java.lang.Integer, java.lang.String>
        objectB = objectB -- com.zfoo.protocol.packet.ObjectB
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function ObjectA:protocolId()
    return 102
end

function ObjectA:write(buffer, packet)
    if buffer:writePacketFlag(packet) then
        return
    end
    buffer:writeInt(packet.a)
    buffer:writeIntStringMap(packet.m)
    buffer:writePacket(packet.objectB, 103)
end

function ObjectA:read(buffer)
    if not(buffer:readBoolean()) then
        return nil
    end
    local packet = ObjectA:new()
    local result0 = buffer:readInt()
    packet.a = result0
    local map1 = buffer:readIntStringMap()
    packet.m = map1
    local result2 = buffer:readPacket(103)
    packet.objectB = result2
    return packet
end

return ObjectA
